PhotoBooth/Components/Capture Session/CaptureSession.swift

```diff
@@ -53,12 +74,70 @@ class CaptureSession: NSObject, PhotoCaptureable, AVCapturePhotoCaptureDelegate
         setupPhotoCaptureSession()
         setupDevices()
         setUpCaptureSessionInput(position: .front)
+        setFlashMode(mode: .off)
     }
 
     func captureImage(){
         takePhoto()
     }
 
+    func getFlashMode() -> AVCaptureDevice.FlashMode {
+        return currentFlashMode!
+    }
+
+    func setFlashMode(mode: FlashModeOption) {
+        guard let camera = currentCamera else { return }
+        if camera.hasFlash {
+            switch mode {
+            case .on:
+                currentFlashMode = .on
+            case .auto:
+                currentFlashMode = .auto
+            default:
+                currentFlashMode = .off
+            }
+        }
+    }
+
+    func focus(touchLocation: CGPoint) {
+        let focusPoint = touchLocation
+
+        if let device = currentCamera {
+            do {
+                try device.lockForConfiguration()
+                if device.isFocusPointOfInterestSupported {
+                    device.focusPointOfInterest = focusPoint
+                    device.focusMode = AVCaptureDevice.FocusMode.autoFocus
+                }
+                if device.isExposurePointOfInterestSupported {
+                    device.exposurePointOfInterest = focusPoint
+                    device.exposureMode = AVCaptureDevice.ExposureMode.autoExpose
+                }
+                device.unlockForConfiguration()
+            } catch {
+                // Handle errors here
+                print("\(error.localizedDescription)")
+            }
+        }
+    }
+
+    // Return zoom value between the minimum and maximum zoom values
+    func minMaxZoom(_ factor: CGFloat) -> CGFloat {
+        return min(min(max(factor, CameraZoom.min), CameraZoom.max), currentCamera!.activeFormat.videoMaxZoomFactor)
+    }
+
+    func updateZoomScaleFactor(scale factor: CGFloat) {
+        if let device = currentCamera {
+            do {
+                try device.lockForConfiguration()
+                defer { device.unlockForConfiguration() }
```

Comment:
you understand what defer is doing?

Reply:
as far as I understand the defer statement will always run last in the function

Reply:
yep, just making sure it was intentional